projects/crash_ping_delays.kp/knowledge.md

```diff
@@ -0,0 +1,252 @@
+
+**Recording Delay** is the time from when the data "happens" to the time we record it in a ping. 
+
+Due to only having day-resolution time information about the crash, this will be approximate and might look weird.
+
+
+```python
+setup_plot("Recording Delay CDF", MAX_DELAY_S / HOUR_IN_S)
+
+for chan in CHANNELS:
+    plot_cdf(delays_by_chan\
+             .filter(lambda d: d[0] == chan)\
+             .map(lambda d: d[1][0] / HOUR_IN_S if d[1][0] < MAX_DELAY_S else MAX_DELAY_S / HOUR_IN_S)\
+             .collect())
+    
+plt.legend(CHANNELS, loc="lower right")
+```
+
+
+
+
+    <matplotlib.legend.Legend at 0x7f8dcb4c8810>
+
+
+
+
+
+![png](images/output_19_1.png)
+
+
+Since we don't know when in the day a crash happened, we can't use the precise time of day the ping was created to tell us how long it's been. Thus we get this stair-step pattern as each ping is some quantum of days.
+
+Still, it's enough to show us that Nightly is a clear winner with over 95% of its crashes recorded within a day. Release and beta still manage over 70% within a day and over 80% within two. However, it takes at least four days to reach 90%.
+
+### Submission Delay
+
+**Submission Delay** is the delay between the data being recorded on the client and it being received by our infrastructure. It is thought to be dominated by the length of time Firefox isn't open on a client's computer, though retransmission attempts and throttling can also contribute.
+
+Here we run into a problem with **clock skew**. Clients' clocks aren't guaranteed to align with our server's clock, so we cannot necessarily compare the two. Luckily, with [bug 1144778](https://bugzilla.mozilla.org/show_bug.cgi?id=1144778) we introduced an HTTP `Date` header which tells us what time the client's clock thinks it is when it is sending the data. Coupled with the `Timestamp` field recorded which is what time the server's clock thinks it is when it receives the data, we can subtract the more egregious examples of clock skew and get values that are closer to reality.
+
+
+```python
+setup_plot("Submission Delay CDF", MAX_DELAY_S / HOUR_IN_S)
+
+for chan in CHANNELS:
+    plot_cdf(delays_by_chan\
+             .filter(lambda d: d[0] == chan)\
+             .map(lambda d: d[1][1] / HOUR_IN_S if d[1][1] < MAX_DELAY_S else MAX_DELAY_S / HOUR_IN_S)\
+             .collect())
+    
+plt.legend(CHANNELS, loc="lower right")
+```
+
+
+
+
+    <matplotlib.legend.Legend at 0x7f8dbe677a10>
+
+
+
+
+
+![png](images/output_22_1.png)
+
+
+Submission delays are _really_ low across the board meaning there is very little time between the crash ping being created and it being received by our servers.
+
+This echoes the code where the creation of the crash ping happens on the next restart of the browser, and is then sent almost immediately.
+
+Nightly is an interesting subject, though, in that it starts out as the slowest performer before becoming the channel with the most submitted crashes after 24 hours.
```

Comment:
Submission delay's a weird one as it should be almost universally 0. You restart your browser after a short (Nightly) or longer (Release) recording delay and only then is the crash ping created. But by the time it's created, gecko's up and running and should be ready to send off the first batch of last session's pings, which _ought_ to include the crashping.

That we see any appreciable number of crash pings _not_ being sent immediately is a curiosity. Because if they were never sent, they wouldn't be counted on the plot. But there clearly are some pings that we receive over _four days_ after the ping was created (5% of release crash pings aren't in under 96 hours. Five!). After at least four days the user sends them to us! How?!

It's a mystery to me. I welcome any speculation at this point.